### PR TITLE
Change Vocab and Trainers to prepare non-subword training. [Idx enum]

### PR DESCRIPTION
**Pull request by sebpuetz**
First commit to allow training without subwords. Adds the Idx enum representing a Single or Multiple indices.

I think this is cleaner than the original proposition in #13, there are probably still a few things to discuss, but I hope this is going in the right direction.

### CODE REVIEW

File: finalfrontier/src/vocab.rs [hunk 1]
```diff
@@ -213,10 +213,29 @@ impl From<SubwordVocab> for VocabWrap {
         .into()
     }
 }
+/// Index representation.
+///
+/// This enum represents possible types of indices, e.g. in a vocabulary with subwords, the `Multi`
+/// variant includes all indices associated with a key.
+#[derive(Clone)]
+pub enum Idx {
+    Single(usize),
```
Comment (sebpuetz) [inline finalfrontier/src/vocab.rs]: Should this also be `u64` or should both just be `usize`, what's the motivation for `u64` in the indexing methods?

Comment (danieldk) [reply finalfrontier/src/vocab.rs]: The `u64` comes from hashing in subword indices, since we use a 64-bit hash. I have to think a bit about this. In Rust-land, there are no obvious benefits, since on a 32-bit machine we have to index on arrays at some point using usize and it is unlikely that we'll have more than 2^32 embedding array rows anytime soon. However, `u64` is nice for a C ABIs because `usize` does not have an equivalent integer type in C outside `uintptr_t`, whereas `u64` is equivalent to `uint64_t`.

File: finalfrontier/src/vocab.rs [hunk 2]
```diff
@@ -213,10 +213,29 @@ impl From<SubwordVocab> for VocabWrap {
         .into()
     }
 }
+/// Index representation.
+///
+/// This enum represents possible types of indices, e.g. in a vocabulary with subwords, the `Multi`
+/// variant includes all indices associated with a key.
+#[derive(Clone)]
+pub enum Idx {
```
Comment (danieldk) [inline finalfrontier/src/vocab.rs]: This approach has the downside that without subwords, indices are still ptr + cap + len size due to the largest enum variant (+ possible padding / alignment).

**Comment by sebpuetz** [comment]
Sure, there's no need to rush this. I might look into an implementation with associated types. I might be able to work around the memory overhead. But anyways, memory should only be wasted in case of the no-subwords training. Training with subword indices shouldn't suffer from any issues here, which I think should also be the focus.

File: finalfrontier/src/sgd.rs [hunk 3]
```diff
@@ -152,9 +151,17 @@ impl NegativeSamplingSGD {
         loss += self.negative_samples(model, input_embed, input_delta.view_mut(), output, lr);
 
         // Update the input embeddings with the accumulated gradient.
-        for &idx in input {
-            let input_embed = model.input_embedding_mut(idx as usize);
-            scaled_add(input_embed, input_delta.view(), 1.0);
+        match input {
+            Idx::Single(idx) => {
```
Comment (danieldk) [inline finalfrontier/src/sgd.rs]: Myabe this could be shortened if Idx was iterable? The both cases could be combined into one case that that iterates over an idx instance.

File: finalfrontier/src/skipgram_trainer.rs [hunk 4]
```diff
@@ -211,13 +211,15 @@ where
             let right = cmp::min(self.i + context_size + 1, self.ids.len());
             let contexts = (left..right)
                 .filter(|&idx| idx != self.i)
-                .map(|idx| self.output_(self.ids[idx], self.i, idx))
+                .map(|idx| self.output_(self.ids[idx].single_idx() as usize, self.i, idx))
                 .fold(Vec::with_capacity(right - left), |mut contexts, idx| {
                     contexts.push(idx);
                     contexts
                 });
+            let mut word_idx = Idx::Single(self.ids[self.i].single_idx());
```
Comment (danieldk) [inline finalfrontier/src/skipgram_trainer.rs]: This is hard to follow... Can't the iterator just continue to use `usize` for the focus token? Maybe just map the `Vec` to one of `usize` in the constructor?

File: finalfrontier/src/train_model.rs [hunk 5]
```diff
@@ -101,8 +101,11 @@ impl<T> TrainModel<T> {
     }
 
     /// Get the mean input embedding of the given indices.
-    pub(crate) fn mean_input_embedding(&self, indices: &[u64]) -> Array1<f32> {
-        Self::mean_embedding(self.input.view(), indices)
+    pub(crate) fn mean_input_embedding(&self, idx: &Idx) -> Array1<f32> {
+        match idx {
+            Idx::Single(idx) => self.input.subview(Axis(0), *idx).to_owned(),
```
Comment (danieldk) [inline finalfrontier/src/train_model.rs]: `.row`

File: finalfrontier/src/vocab.rs (hunk 1 above)
Comment (danieldk) [reply finalfrontier/src/vocab.rs]: Both enum variants should have wrap the same type. So, I guess both `usize` for now.

File: finalfrontier/src/vocab.rs [hunk 6]
```diff
@@ -213,10 +213,29 @@ impl From<SubwordVocab> for VocabWrap {
...
+    Single(usize),
+    Multi(Vec<u64>),
+}
+
+impl Idx {
+    /// Return the unique index for the index.
+    pub fn single_idx(&self) -> usize {
+        match self {
+            Idx::Single(idx) => *idx,
+            Idx::Multi(v) => *v.last().unwrap() as usize,
```
Comment (danieldk) [inline finalfrontier/src/vocab.rs]: Now this assumes that the word index is always the last index? This seems quite brittle. I know that this has the convenience of being able to provide a `&[u64]` in various places, but it would be nice to have stronger invariants.

File: finalfrontier/src/train_model.rs (hunk 5 above)
Comment (sebpuetz) [reply finalfrontier/src/train_model.rs]: We should then fix the other calls in `train_model`, too. The line was just copy pasted iirc.

File: finalfrontier/src/vocab.rs (hunk 6 above)
Comment (sebpuetz) [reply finalfrontier/src/vocab.rs]: Could probably made a tuple `(Vec<usize>, usize)` to get the single focus idx.

File: finalfrontier/src/skipgram_trainer.rs (hunk 4 above)
Comment (sebpuetz) [reply finalfrontier/src/skipgram_trainer.rs]: Do you mean `(usize, Vec<usize>` as `Item` for the iterator?

I'm not sure what you mean by mapping `Vec` and `usize`, get the single index from the indices `Vec` in the constructor and map those indices to the vectors?

Comment (sebpuetz) [reply finalfrontier/src/skipgram_trainer.rs]: I went for the `mem::swap` and construction of a single idx in both proposed PRs because we can't move out of `self.ids` and we don't get `Copy` on the `Idx` types. So there would be a few options I guess:

* Store two id `Vec`s in the struct, one for the focus indices and one for the context indices.
* Store a single id `Vec`, don't touch the entries and `.clone()` the indices to return them.
* The proposed solution, construct a cheap alternative and swap it with the focus indices we need to return.

I think if we don't want to abstract over the index types in this struct, then we'd need to add an explicit indirection in `Trainer`, which was a solution I proposed quite some time ago with some `AssociatedIndices` trait that returned an empty slice for `SimpleVocab` and a slice with subword indices for `SubwordVocab`.

**Comment by sebpuetz** [comment]
Turns out I actually had a rather clean implementation with an enum in #13 and forgot about it. I rebased master on the branch and reopened the PR.

File: finalfrontier/src/skipgram_trainer.rs (hunk 4 above)
Comment (danieldk) [reply finalfrontier/src/skipgram_trainer.rs]: I don't mean the iterator item, but the member `ids` member, which used to be `Vec<usize>` and is now `Vec<Idx>`. I thought the constructor could just `map` + `collect` the `Vec<Idx>` to `Vec<usize>` and use that. But I see now that the iterator-using code now uses `Multi`.

Comment (danieldk) [reply finalfrontier/src/skipgram_trainer.rs]: I think that I now understand better why the code is difficult to read in places. `Idx` is very generic and `single_idx` does not really convey what is meant.

I think you'd rather want some name like `WordIndex` and methods could be `word_idx` and `subword_indices` (if it is the variant with subwords). But I will read the trait-based PR, because this really sounds like it should just be an associated type of the vocab.

In this particular case, a code comment describing what happens helps.

File: finalfrontier/src/vocab.rs (hunk 6 above)
Comment (danieldk) [reply finalfrontier/src/vocab.rs]: I have looked through all PRs again, and I think that this is the clearest iteration, though it was still very useful to see the trait version.

I think this is ready to go with two modifications:

1. Changing `Multi` in a tuple as you suggested.
2. Renaming of `Idx` to `WordIdx`, with the methods `word_idx` (applicable to both enum variants) and `subword_indices` (only applicable to `Multi`). I would also consider changing `Single` to `Word` and `Multi` to `WordWithSubwords` (or something along those lines).

The variant renaming would also make it possible to have a `Subwords` variant later with only subwords.

Comment (sebpuetz) [reply finalfrontier/src/vocab.rs]: Thanks for the feedback! I'm not really happy with any of the 2 (and a half) iterations I proposed so far. I just pushed the cleaned up version anyways.

> The variant renaming would also make it possible to have a `Subwords` variant later with only subwords.

Although implementing `word_idx` on such a variant would be odd.
